hashintel/hash
# Properly express the Turborepo dependency graph

## 🌟 What is the purpose of this PR?

During the process of setting up Turborepo, we reprioritised work and the repo was left in an inconsistent state. This has lead to a number of interim patches such as disabling caching in CI.

This PR attempts to correctly express dependencies between packages/workspaces within the repository. This was made complicated by a few things
- We have Rust applications/libraries which interact with TypeScript
- The `package.json` setups of blocks are massively inconsistent and involved a fair amount of manual work.
- Our `codegen` step is sometimes used to produce artifacts for normal usage of the package (e.g. running in development, linting, building), and sometimes used to produce artifacts for downstream consumptions of the package. This makes it difficult to express it at the top of the repo with simplistic definitions like "build depends on codegen"
- Our `build` step is somewhat inconsistently used to mean "be ready for production usage" or "be ready for any usage at all", and this affects how we want dependencies to be expressed, especially with the extremely common "^build" expression. We do not want to be running full production builds of applications if we are importing elements of their API surface for example (particularly relevant with the `status` package)

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1204000740778935/1203641339878924/f) _(internal)_

## 🚫 Blocked by

- N/A

## 🔍 What does this change?

- Modifies the top-level `turbo.json` to be more general
- Adds `turbo.json` files to packages that require custom script definitions
- A few drive-by cleanups

## 🔍 Has this modified a publishable library?




This PR:


## 📜 Does this require a change to the docs?

- We should probably invest some effort into describing the monorepo structure adequately in repo documentation. As we do not have that, there are no changes required right now.

## ⚠️ Known issues

- We inconsistently describe codegen steps (especially for blocks, with `schema` or `codegen`, this is a known issue due to out of date dependencies)
- The `build` semantics inconsistency described above
- A number of blocks are missing `eslint`, `tsc`, or `format` scripts
- The semantics of some scripts have been changed (previously some blocks had `codegen` calling `format` as part of the script). We perhaps want to revert some of those changes. I have made the changes to better represent the dependencies in the turbo graph, but if people `cd` into the package directories and simply run the scripts (for example, `yarn codegen` within a block directory) then they will not benefit from turborepo's dependency graph. We will need to use `turbo run` for this.
  - _Edit: this was reverted in [Revert removal of format from codegen command in blocks](https://github.com/hashintel/hash/pull/2374/commits/cc319247da67c3d43d5dd22488fe55f71eb9230b) as it caused problems where running `yarn` would result in a git diff because generated files weren't being formatted, and we don't want to reformat the repo on every `yarn` run_

## 🐾 Next steps

Unknown at this time

## 🛡 What tests cover this?

This should mostly impact CI, all linting, tests, builds, etc. are important to be checked

## ❓ How to test this?

1. View CI
2. Try scripts from the top level (e.g. `yarn install`)
3. Try individual workspaces such as blocks by running with `turbo run`
4. Try building the dependency graph with `turbo run build --graph=graph.svg` (you can change the line ending to `.pdf` or `.dot` for other formats)
  - The root dependency graph is pretty unreadable (see demo), you can check the dependency graph of individual packages with a filter, e.g.: `turbo run build --filter=@apps/hash-api --graph=graph.svg` 

## 📹 Demo

![graph](https://user-images.githubusercontent.com/25749103/231485248-4778996c-d945-4f34-973b-421277eca4ae.svg)
